Add reference to the user-account.user_id from process.creator_user_ref for qradar and elastic_ecs modules

aviv1ron1 opened the pull request:
In the stix 2.0 definition under the `process` SCO there is an optional field for `creator_user_ref`. According to the spec this field "Specifies the user that created the process, as a reference to a User Account Object."
This PR adds this reference to the process  object for the `qradar` and `elastic_ecs` modules.
It maps it to the `username` field in `qradar` and to the `user.name` field in `elastic_ecs`.
Unit tests have been updated accordingly to pass.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/opencybersecurityalliance/stix-shifter?pullRequest=462) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/opencybersecurityalliance/stix-shifter?pullRequest=462) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/opencybersecurityalliance/stix-shifter?pullRequest=462) it.</sub>

mdazam1942 commented on stix_shifter_modules/qradar/tests/stix_translation/test_qradar_json_to_stix.py:
```diff
@@ -135,11 +135,17 @@ def test_cybox_observables(self):
+        assert(proc_obj is not None), 'process object type not found'
+        assert(proc_obj.keys() == {'type', 'creator_user_ref'})
+        user_ref = proc_obj['creator_user_ref']
+        assert(user_ref in objects), f"creator_user_ref with key {proc_obj['creator_user_ref']} not found"
```
an unittets that represents process object that reference username would be ideal when we add new mapping of such instead of `not found`

aviv1ron1 replied:
Thats what the updated unit test does - it checks that process object exists, and that it has a valid reference to the user. Am I missing something?

mdazam1942 replied:
I'm confused. you don't have any process object but you are still creating one with user reference. not sure what you are trying to map.

mdazam1942 replied:
```
{"sourceip": source_ip, "destinationip": destination_ip, "url": url, "eventpayload": payload, "username": user_id, "protocol": 'TCP', "sourceport": "3000", "destinationport": 2000, "filename": file_name, "domainname": domain, "sourcemac": source_mac, "destinationmac": destination_mac}
```
this is the json results that it is trying to translate with your new mapping,

mdazam1942 commented on stix_shifter_modules/qradar/stix_translation/json/to_stix_map.json:
```diff
@@ -4,9 +4,17 @@
+    {
+      "key": "process.creator_user_ref",
+      "object": "process_image",
+      "references": "user"
```
nothing wrong with the mapping. but its not making any sense. there's no process object mapping and yet user reference is created.

aviv1ron1 replied:
makes sense, this is because the rest of the process fields are still missing for qradar - I will add them to this PR

aviv1ron1 commented:
elastic_ecs module: 
added a reference from the process object to the user object

QRadar module: 
Added the process object.
Before: 

```
{
    "0":
    {
        "type": "ipv6-addr",
        "value": "fd80:655e:171d:30d4:fd80:655e:171d:30d4",
        "resolves_to_refs": ["10"]
    },
    "1":
    {
        "type": "network-traffic",
        "src_ref": "0",
        "dst_ref": "3",
        "protocols": ["tcp"],
        "src_port": 3000,
        "dst_port": 2000
    },
    "2":
    {
        "type": "x-ibm-finding",
        "src_ip_ref": "0",
        "dst_ip_ref": "3",
        "src_application_user_ref": "7"
    },
    "3":
    {
        "type": "ipv4-addr",
        "value": "255.255.255.1",
        "resolves_to_refs": ["11"]
    },
    "4":
    {
        "type": "url",
        "value": "https://example.com"
    },
    "5":
    {
        "type": "domain-name",
        "value": "example.com"
    },
    "6":
    {
        "type": "artifact",
        "payload_bin": "dXRmIHBheWxvYWQ="
    },
    "7":
    {
        "type": "user-account",
        "user_id": "someuserid2018"
    },
    "8":
    {
        "type": "file",
        "name": "somefile.exe"
    },
    "9":
    {
        "type": "domain-name",
        "value": "test.com"
    },
    "10":
    {
        "type": "mac-addr",
        "value": "00-00-5E-00-53-00"
    },
    "11":
    {
        "type": "mac-addr",
        "value": "00-00-5A-00-55-01"
    }
}
```

after (look at object number 8 and its references):
```
{
    "0":
    {
        "type": "ipv6-addr",
        "value": "fd80:655e:171d:30d4:fd80:655e:171d:30d4",
        "resolves_to_refs": ["11"]
    },
    "1":
    {
        "type": "network-traffic",
        "src_ref": "0",
        "dst_ref": "3",
        "protocols": ["tcp"],
        "src_port": 3000,
        "dst_port": 2000
    },
    "2":
    {
        "type": "x-ibm-finding",
        "src_ip_ref": "0",
        "dst_ip_ref": "3",
        "src_application_user_ref": "7"
    },
    "3":
    {
        "type": "ipv4-addr",
        "value": "255.255.255.1",
        "resolves_to_refs": ["12"]
    },
    "4":
    {
        "type": "url",
        "value": "https://example.com"
    },
    "5":
    {
        "type": "domain-name",
        "value": "example.com"
    },
    "6":
    {
        "type": "artifact",
        "payload_bin": "dXRmIHBheWxvYWQ="
    },
    "7":
    {
        "type": "user-account",
        "user_id": "someuserid2018"
    },
    "8":
    {
        "type": "process",
        "creator_user_ref": "7",
        "binary_ref": "13",
        "parent_ref": "17",
        "command_line": "C:\\example\\executable.exe --example",
        "extensions":
        {
            "windows-service-ext":
            {
                "service_dll_refs[*]":
                {
                    "name": "C:\\example\\some.dll"
                }
            }
        }
    },
    "9":
    {
        "type": "file",
        "name": "somefile.exe"
    },
    "10":
    {
        "type": "domain-name",
        "value": "test.com"
    },
    "11":
    {
        "type": "mac-addr",
        "value": "00-00-5E-00-53-00"
    },
    "12":
    {
        "type": "mac-addr",
        "value": "00-00-5A-00-55-01"
    },
    "13":
    {
        "type": "file",
        "name": "proc_img.exe",
        "parent_directory_ref": "14"
    },
    "14":
    {
        "type": "directory",
        "path": "C:\\example\\process\\image"
    },
    "15":
    {
        "type": "file",
        "name": "proc_parent_img.exe",
        "parent_directory_ref": "16"
    },
    "16":
    {
        "type": "directory",
        "path": "C:\\example\\process\\parent\\image"
    },
    "17":
    {
        "type": "process",
        "binary_ref": "15",
        "command_line": "C:\\example\\parent.exe --example"
    }
}
```

To get relevant process data it is required to install the `BM QRadar Content Extension for Sysmon` content pack for QRadar.
https://exchange.xforce.ibmcloud.com/hub/extension/e41e758e2ab5786173438cd09219a9d0

mdazam1942 commented:
> elastic_ecs module:
> added a reference from the process object to the user object
> 
> QRadar module:
> Added the process object.
> Before:
> 
> ```
> {
>     "0":
>     {
>         "type": "ipv6-addr",
>         "value": "fd80:655e:171d:30d4:fd80:655e:171d:30d4",
>         "resolves_to_refs": ["10"]
>     },
>     "1":
>     {
>         "type": "network-traffic",
>         "src_ref": "0",
>         "dst_ref": "3",
>         "protocols": ["tcp"],
>         "src_port": 3000,
>         "dst_port": 2000
>     },
>     "2":
>     {
>         "type": "x-ibm-finding",
>         "src_ip_ref": "0",
>         "dst_ip_ref": "3",
>         "src_application_user_ref": "7"
>     },
>     "3":
>     {
>         "type": "ipv4-addr",
>         "value": "255.255.255.1",
>         "resolves_to_refs": ["11"]
>     },
>     "4":
>     {
>         "type": "url",
>         "value": "https://example.com"
>     },
>     "5":
>     {
>         "type": "domain-name",
>         "value": "example.com"
>     },
>     "6":
>     {
>         "type": "artifact",
>         "payload_bin": "dXRmIHBheWxvYWQ="
>     },
>     "7":
>     {
>         "type": "user-account",
>         "user_id": "someuserid2018"
>     },
>     "8":
>     {
>         "type": "file",
>         "name": "somefile.exe"
>     },
>     "9":
>     {
>         "type": "domain-name",
>         "value": "test.com"
>     },
>     "10":
>     {
>         "type": "mac-addr",
>         "value": "00-00-5E-00-53-00"
>     },
>     "11":
>     {
>         "type": "mac-addr",
>         "value": "00-00-5A-00-55-01"
>     }
> }
> ```
> 
> after (look at object number 8 and its references):
> 
> ```
> {
>     "0":
>     {
>         "type": "ipv6-addr",
>         "value": "fd80:655e:171d:30d4:fd80:655e:171d:30d4",
>         "resolves_to_refs": ["11"]
>     },
>     "1":
>     {
>         "type": "network-traffic",
>         "src_ref": "0",
>         "dst_ref": "3",
>         "protocols": ["tcp"],
>         "src_port": 3000,
>         "dst_port": 2000
>     },
>     "2":
>     {
>         "type": "x-ibm-finding",
>         "src_ip_ref": "0",
>         "dst_ip_ref": "3",
>         "src_application_user_ref": "7"
>     },
>     "3":
>     {
>         "type": "ipv4-addr",
>         "value": "255.255.255.1",
>         "resolves_to_refs": ["12"]
>     },
>     "4":
>     {
>         "type": "url",
>         "value": "https://example.com"
>     },
>     "5":
>     {
>         "type": "domain-name",
>         "value": "example.com"
>     },
>     "6":
>     {
>         "type": "artifact",
>         "payload_bin": "dXRmIHBheWxvYWQ="
>     },
>     "7":
>     {
>         "type": "user-account",
>         "user_id": "someuserid2018"
>     },
>     "8":
>     {
>         "type": "process",
>         "creator_user_ref": "7",
>         "binary_ref": "13",
>         "parent_ref": "17",
>         "command_line": "C:\\example\\executable.exe --example",
>         "extensions":
>         {
>             "windows-service-ext":
>             {
>                 "service_dll_refs[*]":
>                 {
>                     "name": "C:\\example\\some.dll"
>                 }
>             }
>         }
>     },
>     "9":
>     {
>         "type": "file",
>         "name": "somefile.exe"
>     },
>     "10":
>     {
>         "type": "domain-name",
>         "value": "test.com"
>     },
>     "11":
>     {
>         "type": "mac-addr",
>         "value": "00-00-5E-00-53-00"
>     },
>     "12":
>     {
>         "type": "mac-addr",
>         "value": "00-00-5A-00-55-01"
>     },
>     "13":
>     {
>         "type": "file",
>         "name": "proc_img.exe",
>         "parent_directory_ref": "14"
>     },
>     "14":
>     {
>         "type": "directory",
>         "path": "C:\\example\\process\\image"
>     },
>     "15":
>     {
>         "type": "file",
>         "name": "proc_parent_img.exe",
>         "parent_directory_ref": "16"
>     },
>     "16":
>     {
>         "type": "directory",
>         "path": "C:\\example\\process\\parent\\image"
>     },
>     "17":
>     {
>         "type": "process",
>         "binary_ref": "15",
>         "command_line": "C:\\example\\parent.exe --example"
>     }
> }
> ```
> 
> To get relevant process data it is required to install the `BM QRadar Content Extension for Sysmon` content pack for QRadar.
> https://exchange.xforce.ibmcloud.com/hub/extension/e41e758e2ab5786173438cd09219a9d0

mdazam1942 commented:
@aviv1ron1 if process mappings only works with a particular QRadar content pack than we won't be able to put these fields in the default QRadar mappings. That will break the translation for others who doesn't have that content pack installed in QRadar datasource. 

I had the impression aql may have some process fields by default that can be queried. But looks like it doesn't have any process related fields that can be used in the query. Mappings in the from/to stix files may not break the translations because if nothing is return for those fields than those won't have any effect. But putting those fields in `aql_events_fields.json` breaks the aql query if those are not present in the QRadar datasource.

@delliott90 wondering if you know any other way to accommodate these process fields?

mdazam1942 commented on stix_shifter_modules/qradar/stix_translation/json/aql_events_fields.json:
```diff
@@ -41,6 +41,31 @@
-    "DOMAINNAME(domainid) as domainname"
+    "DOMAINNAME(domainid) as domainname",
+    "EventID",
+    "Image",
```
by default events table in aql doesn't have any of these fields that you added. which throws invalid aql query exception.

aviv1ron1 commented:
Cancelling this pull request - I will submit a new one with all the changes in one PR
